libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp

```diff
@@ -600,6 +606,58 @@ namespace hpx::execution::experimental {
     template <typename Sender, typename Env = empty_env>
     inline constexpr bool is_sender_v = is_sender<Sender, Env>::value;
 
+    namespace detail {
+
+        // Sender may or may not support sends_stopped
+        // In the false case only the value_types and error_types are checked
+        template <class S, class R, class CS,
+            bool sends_stopped = CS::sends_stopped>
+        struct sender_support_sends_stopped;
+
+        template <class S, class R, class CS>
+        struct sender_support_sends_stopped<S, R, CS, true>
+          : std::integral_constant<bool,
+                is_receiver_of_v<R, completion_signatures_of_t<S, env_of_t<R>>>>
+        {
+        };
+
+        template <class S, class R, class CS>
+        struct sender_support_sends_stopped<S, R, CS, false>
+          : std::integral_constant<bool,
+                is_invocable_variant_of_tuples<set_value_t, R,
+                    typename CS::template value_types<hpx::tuple,
+                        hpx::variant>>::value &&
+                    is_invocable_variant<set_error_t, R,
+                        typename CS::template error_types<hpx::variant>>::value>
+        {
+        };
+
+        template <bool IsSenderOf, class S, class R>
+        struct is_sender_of_impl;
+
+        template <class S, class R>
+        struct is_sender_of_impl<false, S, R> : std::false_type
+        {
+        };
+
+        template <class S, class R>
+        struct is_sender_of_impl<true, S, R>
+          : std::integral_constant<bool,
+                sender_support_sends_stopped<S, R,
+                    completion_signatures_of_t<S, env_of_t<R>>>::value>
+        {
+        };
+    }    // namespace detail
+
+    template <class S, class R>
+    struct is_sender_of
+      : detail::is_sender_of_impl<is_sender_v<S, env_of_t<R>>, S, R>
+    {
+    };
```

Comment:
Apparently, we have overdone this. P2300 defines `sender_of` as:
```
template<class S, class E = no_env, class... Ts>
  concept sender_of =
    sender<S, E> &&
    same_as<
      type-list<Ts...>,
      value_types_of_t<S, E, type-list, type_identity_t>
    >;
```
IOW, all we need to check is whether the possible arguments to `set_value()` match the given list of arguments. The confusion was caused (well, my confusion) by the fact that there are two different concepts: `sender_of` and `sender_to`. The first checks the value-types only while the second verifies that `connect()` can be called on a particular pair of a sender and a receiver:
```
template<class S, class R>
  concept sender_to =
    sender<S, env_of_t<R>> &&
    receiver_of<R, completion_signatures_of_t<S, env_of_t<R>>> &&
    requires (S&& s, R&& r) {
      execution::connect(std::forward<S>(s), std::forward<R>(r));
    };
```

To make the picture complete, there is a `receiver_of` concept that verifies whether a given receiver type can handle the given completion signatures:
```
template <class T, class Completions>
  concept receiver_of =
    receiver<T> &&
    requires (Completions* completions) {
        []<valid-completion-for<T>...Sigs>(completion_signatures<Sigs...>*)
        {}(completions);
    };
```

I apologize for the confusion but it seems you will have to revisit your implementation of `is_sender_of` one more time.

Reply:
Hi, thanks for clarifying, I was going to do it https://github.com/STEllAR-GROUP/hpx/commit/86cca5e08514ce457bbdab472a51b728cadfa4f3 but then I thought maybe it's supposed to be this way. I could have mentioned it to you, sorry.
Let me redo this and also think over the other comments.
Thanks

Reply:
> IOW, all we need to check is whether the possible arguments to set_value() match the given list of arguments.

IOW, we need to check if set_value is invocable with the given list of arguments?

Reply:
If it is invocable then it implies that the types match, is this implementation okay or I should go by with some other mechanism?